### PR TITLE
OSDOCS-8661: Configuring IP failover content audit

### PR DESCRIPTION
**Pull request by ahardin-rh**
Version(s): 4.12+

Issue:
https://issues.redhat.com/browse/OSDOCS-8661

Link to docs preview:
https://74064--ocpdocs-pr.netlify.app/openshift-enterprise/latest/networking/configuring-ipfailover.html

QE review:
- [x] QE has approved this change.

### CODE REVIEW

File: modules/nw-ipfailover-cluster-ha-ingress.adoc [hunk 1]
```diff
@@ -3,10 +3,10 @@
...
+= High availability For ExternalIP
 
-In non-cloud clusters, IP failover and `ingressIP` to a service can be combined. The result is high availability services for users that create services using `ingressIP`.
+In non-cloud clusters, IP failover and `ExternalIP` to a service can be combined. The result is high availability services for users that create services using `ExternalIP`.
 
-The approach is to specify an `ingressIPNetworkCIDR` range and then use the same range in creating the ipfailover configuration.
+The approach is to specify an `spec.ExternalIP.autoAssignCIDRs` range and then use the same range in creating the IP failover configuration.
 
-Because IP failover can support up to a maximum of 255 VIPs for the entire cluster, the `ingressIPNetworkCIDR` needs to be `/24` or smaller.
+Because IP failover can support up to a maximum of 255 VIPs for the entire cluster, the `spec.ExternalIP.autoAssignCIDRs` must be `/24` or smaller.
```
Comment (rfredette) [inline modules/nw-ipfailover-cluster-ha-ingress.adoc:12]: I think it'd be good to clarify that `spec.ExternalIP.autoAssignCIDRs` is a field of the cluster network config. If I was reading this without additional context, I'd think this was referring to the ip failover deployment's `spec`.

I'm not sure if there's a style guideline for referring to other cluster resources, but the relevant resource is `network.config/cluster`, and some info on it is [here](https://docs.openshift.com/container-platform/4.15/networking/cluster-network-operator.html#nw-cno-view_cluster-network-operator), if that's helpful.

**Comment by ahardin-rh** [comment]
@melvinjoseph86 Can you please provide QE ack for these changes? Thanks!

File: modules/nw-ipfailover-configuring-check-notify-scripts.adoc [hunk 2]
```diff
@@ -6,26 +6,27 @@
 [id="nw-ipfailover-configuring-check-notify-scripts_{context}"]
 = Configuring check and notify scripts
 
-Keepalived monitors the health of the application by periodically running an optional user supplied check script. For example, the script can test a web server by issuing a request and verifying the response.
+Keepalived monitors the health of the application by periodically running an optional user-supplied check script. For example, the script can test a web server by issuing a request and verifying the response. As cluster administrator, you can provide an optional notify script, which is called whenever the state changes.
+The check and notify scripts run in the IP failover pod and use the pod file system, not the host file system. However, the IP failover pod makes the host file system available under the `/hosts` mount path. When configuring a check or notify script, you must provide the full path to the script. The recommended approach for providing the scripts is to use a config map.
```
Comment (ShaunaDiaz) [inline modules/nw-ipfailover-configuring-check-notify-scripts.adoc:10]:
```suggestion

The check and notify scripts run in the IP failover pod and use the pod file system, not the host file system. However, the IP failover pod makes the host file system available under the `/hosts` mount path. When configuring a check or notify script, you must provide the full path to the script. The recommended approach for providing the scripts is to use a configuration map.
```
Second paragraph is tucking up into the first without the line space.

File: modules/nw-ipfailover-configuring-check-notify-scripts.adoc [hunk 3]
```diff
@@ -6,26 +6,27 @@
...
+
+The full path names of the check and notify scripts are added to the Keepalived configuration file, `_/etc/keepalived/keepalived.conf`, which is loaded every time Keepalived starts. The scripts can be added to the pod with a config map as follows.
```
Comment (ShaunaDiaz) [inline modules/nw-ipfailover-configuring-check-notify-scripts.adoc:12]:
```suggestion
The full path names of the check and notify scripts are added to the Keepalived configuration file, `_/etc/keepalived/keepalived.conf`, which is loaded every time Keepalived starts. The scripts can be added to the pod with a configuration map as described in the following methods.
```
Methods might not be the best word here, but we don't end with "as follows" per style guidance.

File: modules/nw-ipfailover-configuring-check-notify-scripts.adoc [hunk 4]
```diff
@@ -6,26 +6,27 @@
...
 
-Each IP failover pod manages a Keepalived daemon that manages one or more virtual IPs (VIP) on the node where the pod is running. The Keepalived daemon keeps the state of each VIP for that node. A particular VIP on a particular node may be in `master`, `backup`, or `fault` state.
+Each IP failover pod manages a Keepalived daemon that manages one or more virtual IP (VIP) addresses on the node where the pod is running. The Keepalived daemon keeps the state of each VIP for that node. A particular VIP on a particular node might be in `master`, `backup`, or `fault` state.
 
-When the check script for that VIP on the node that is in `master` state fails, the VIP on that node enters the `fault` state, which triggers a renegotiation. During renegotiation, all VIPs on a node that are not in the `fault` state participate in deciding which node takes over the VIP. Ultimately, the VIP enters the `master` state on some node, and the VIP stays in the `backup` state on the other nodes.
+If the check script returns non-zero, the node will go into the `backup` state, and any VIPs it holds will be reassigned.
```
Comment (ShaunaDiaz) [inline modules/nw-ipfailover-configuring-check-notify-scripts.adoc:20]:
```suggestion
If the check script returns non-zero, the node enters the `backup` state, and any VIPs it holds are reassigned.
```
No future tense

File: modules/nw-ipfailover-configuring-vrrp-preemption.adoc [hunk 5]
```diff
@@ -6,14 +6,9 @@
 [id="nw-ipfailover-configuring-vrrp-preemption_{context}"]
 = Configuring VRRP preemption
 
-When a Virtual IP (VIP) on a node leaves the `fault` state by passing the check script, the VIP on the node enters the `backup` state if it has lower priority than the VIP on the node that is currently in the `master` state. However, if the VIP on the node that is leaving `fault` state has a higher priority, the preemption strategy determines its role in the cluster.
-
+When a Virtual IP (VIP) on a node leaves the `fault` state by passing the check script, the VIP on the node enters the `backup` state if it has lower priority than the VIP on the node that is currently in the `master` state.
```
Comment (ShaunaDiaz) [inline modules/nw-ipfailover-configuring-vrrp-preemption.adoc:9]:
```suggestion

When a Virtual IP (VIP) on a node leaves the `fault` state by passing the check script, the VIP on the node enters the `backup` state if it has lower priority than the VIP on the node that is currently in the `master` state.
```
looks like we lost a line space

File: modules/nw-ipfailover-remove.adoc [hunk 6]
```diff
@@ -92,10 +92,11 @@ spec:
       - name: remove-ipfailover
         image: quay.io/openshift/origin-keepalived-ipfailover:{product-version}
         command: ["/var/lib/ipfailover/keepalived/remove-failover.sh"]
-      nodeSelector:
+      nodeSelector: <.>
```
Comment (ShaunaDiaz) [inline modules/nw-ipfailover-remove.adoc:95]:
```suggestion
      nodeSelector: <1>
```

File: modules/nw-ipfailover-remove.adoc [hunk 7]
```diff
@@ -92,10 +92,11 @@ spec:
...
         kubernetes.io/hostname: <host_name>  <.>
```
Comment (ShaunaDiaz) [inline modules/nw-ipfailover-remove.adoc:96]:
```suggestion
        kubernetes.io/hostname: <host_name>  <2>
```

File: modules/nw-ipfailover-remove.adoc [hunk 8]
```diff
@@ -92,10 +92,11 @@ spec:
...
       restartPolicy: Never
 ----
+<.> The `nodeSelector` is likely the same as the selector used in the old IP failover deployment.
```
Comment (ShaunaDiaz) [inline modules/nw-ipfailover-remove.adoc:99]:
```suggestion
<1> The `nodeSelector` is likely the same as the selector used in the old IP failover deployment.
```

File: modules/nw-ipfailover-remove.adoc [hunk 9]
```diff
@@ -92,10 +92,11 @@ spec:
...
 <.> Run the job for each node in your cluster that was configured for IP failover and replace the hostname each time.
```
Comment (ShaunaDiaz) [inline modules/nw-ipfailover-remove.adoc:100]:
```suggestion
<2> Run the job for each node in your cluster that was configured for IP failover and replace the hostname each time.
```
https://access.redhat.com/documentation/en-us/openshift_container_platform/4.15/html/networking/configuring-ipfailover#nw-ipfailover-remove_configuring-ipfailover

Existing footnote numbering is already not working on the CP, need to update if adding a second, especially.

File: modules/nw-ipfailover-virtual-ip-addresses-concept.adoc [hunk 10]
```diff
@@ -6,7 +6,7 @@
 [id="nw-ipfailover-virtual-ip-addresses-concept_{context}"]
```
Comment (ShaunaDiaz) [inline modules/nw-ipfailover-virtual-ip-addresses-concept.adoc:6]: It appears this module was removed from the assembly. Is it being used somewhere else? 
https://github.com/openshift/openshift-docs/pull/74064/files#diff-99eebb2772aa83a3592ed7e92101d539b046d16a056823eb8c968ad7b94b0788L55

File: modules/nw-ipfailover-virtual-ip-addresses-concept.adoc [hunk 11]
```diff
@@ -6,7 +6,7 @@
...
 = About virtual IP addresses
 
-Keepalived manages a set of virtual IP addresses (VIP). The administrator must make sure that all of these addresses:
+Keepalived manages a set of virtual IP addresses (VIP). The administrator must ensure that all of these addresses:
```
Comment (ShaunaDiaz) [inline modules/nw-ipfailover-virtual-ip-addresses-concept.adoc:9]:
```suggestion
Keepalived manages a set of virtual IP addresses (VIP). The administrator must ensure that all of the VIP addresses meet the following requirements:
```

File: modules/nw-ipfailover-virtual-ip-addresses-concept.adoc [hunk 12]
```diff
@@ -6,7 +6,7 @@
...
 
 * Are accessible on the configured hosts from outside the cluster.
```
Comment (ShaunaDiaz) [inline modules/nw-ipfailover-virtual-ip-addresses-concept.adoc:11]:
```suggestion
* Accessible on the configured hosts from outside the cluster.
```

File: modules/nw-ipfailover-virtual-ip-addresses-concept.adoc [hunk 13]
```diff
@@ -6,7 +6,7 @@
...
 * Are not used for any other purpose within the cluster.
```
Comment (ShaunaDiaz) [inline modules/nw-ipfailover-virtual-ip-addresses-concept.adoc:12]:
```suggestion
* Not used for any other purpose within the cluster.
```

File: modules/nw-ipfailover-virtual-ip-addresses-concept.adoc [hunk 14]
```diff
@@ -15,5 +15,5 @@ Keepalived on each node determines whether the needed service is running. If it
 
 [NOTE]
 ====
-Each VIP in the set may end up being served by a different node.
+Each VIP in the set might be served by a different node.
```
Comment (ShaunaDiaz) [inline modules/nw-ipfailover-virtual-ip-addresses-concept.adoc:18]:
```suggestion
Each VIP in the set can be served by a different node.
```
Not sure if this is a pre-existing condition that we have to deal with it now ("might be"); or a condition that we are about to potentially create ("can be").

File: networking/configuring-ipfailover.adoc [hunk 15]
```diff
@@ -8,17 +8,24 @@ toc::[]
...
+IP failover uses link:http://www.keepalived.org/[Keepalived] to host a set of externally accessible Virtual IP (VIP) addresses on a set of hosts. Each VIP address is only serviced by a single host at a time. Keepalived uses the Virtual Router Redundancy Protocol (VRRP) to determine which host, from the set of hosts, services which VIP. If a host becomes unavailable, or if the service that Keepalived is watching does not respond, the VIP is switched to another host from the set. This means a VIP is always serviced as long as a host is available.
+
+Every VIP in the set is serviced by a node selected from the set. If a single node is available, the VIPs are served. There is no way to explicitly distribute the VIPs over the nodes, so there can be nodes with no VIPs and other nodes with many VIPs. If there is only one node, all VIPs are on it.
+
+The administrator must ensure that all of the VIP addresses:
```
Comment (ShaunaDiaz) [inline networking/configuring-ipfailover.adoc:15]:
```suggestion
The administrator must ensure that all of the VIP addresses meet the following requirements:
```

File: networking/configuring-ipfailover.adoc [hunk 16]
```diff
@@ -8,17 +8,24 @@ toc::[]
...
+
+* Are accessible on the configured hosts from outside the cluster.
```
Comment (ShaunaDiaz) [inline networking/configuring-ipfailover.adoc:17]:
```suggestion
* Accessible on the configured hosts from outside the cluster.
```

File: networking/configuring-ipfailover.adoc [hunk 17]
```diff
@@ -8,17 +8,24 @@ toc::[]
...
+* Are not used for any other purpose within the cluster.
```
Comment (ShaunaDiaz) [inline networking/configuring-ipfailover.adoc:18]:
```suggestion
* Not used for any other purpose within the cluster.
```

File: networking/configuring-ipfailover.adoc [hunk 18]
```diff
@@ -8,17 +8,24 @@ toc::[]
...
+
+Keepalived on each node determines whether the needed service is running. If it is, VIPs are supported and Keepalived participates in the negotiation to determine which node serves the VIP. For a node to participate, the service must be listening on the watch port on a VIP or the check must be disabled.
 
 [NOTE]
 ====
-The VIPs must be routable from outside the cluster.
+Each VIP in the set might be served by a different node.
```
Comment (ShaunaDiaz) [inline networking/configuring-ipfailover.adoc:24]: same note as previous

File: networking/configuring-ipfailover.adoc [hunk 19]
```diff
@@ -8,17 +8,24 @@ toc::[]
 
 This topic describes configuring IP failover for pods and services on your {product-title} cluster.
```
Comment (ShaunaDiaz) [inline networking/configuring-ipfailover.adoc:10]: "Configuring IP failover" is used twice in this section. Once at the top: https://74064--ocpdocs-pr.netlify.app/openshift-enterprise/latest/networking/configuring-ipfailover.html#nw-ipfailover-configuration_configuring-ipfailover and once again after the environment variables https://74064--ocpdocs-pr.netlify.app/openshift-enterprise/latest/networking/configuring-ipfailover.html#nw-ipfailover-configuration_configuring-ipfailover 

Like this in the live docs: https://access.redhat.com/documentation/en-us/openshift_container_platform/4.15/html/networking/configuring-ipfailover#doc-wrapper

Wondering if you'd be open to re-titling one or the other?

File: networking/configuring-ipfailover.adoc [hunk 20]
```diff
@@ -27,21 +34,16 @@ You can create an IP failover deployment configuration on {product-title}. The I
...
-Setting up a `NodePort` is a privileged operation.
-====
-
 [IMPORTANT]
 ====
 Even though a service VIP is highly available, performance can still be affected. Keepalived makes sure that each of the VIPs is serviced by some node in the configuration, and several VIPs can end up on the same node even when other nodes have none. Strategies that externally load-balance across a set of VIPs can be thwarted when IP failover puts multiple VIPs on the same node.
 ====
 
-When you use `ingressIP`, you can set up IP failover to have the same VIP range as the `ingressIP` range. You can also disable the monitoring port. In this case, all the VIPs appear on same node in the cluster. Any user can set up a service with an `ingressIP` and have it highly available.
+When you use `ExternalIP`, you can set up IP failover to have the same VIP range as the `ExternalIP` range. You can also disable the monitoring port. In this case, all the VIPs appear on same node in the cluster. Any user can set up a service with an `ExternalIP` and have it highly available.
```
Comment (ShaunaDiaz) [inline networking/configuring-ipfailover.adoc:46]:
```suggestion
When you use `ExternalIP`, you can set up IP failover to have the same VIP range as the `ExternalIP` range. You can also disable the monitoring port. In this case, all of the VIPs appear on same node in the cluster. Any user can set up a service with an `ExternalIP` and make it highly available.
```
Is availability configured? Not sure what "have" means in the contexr.

File: networking/configuring-ipfailover.adoc [hunk 21]
```diff
@@ -52,8 +54,6 @@ include::modules/nw-ipfailover-environment-variables.adoc[leveloffset=+1]
 
 include::modules/nw-ipfailover-configuration.adoc[leveloffset=+1]
 
-include::modules/nw-ipfailover-virtual-ip-addresses-concept.adoc[leveloffset=+1]
```
Comment (ShaunaDiaz) [inline networking/configuring-ipfailover.adoc:55]: Here we remove the module edited previously.

File: modules/nw-ipfailover-configuring-check-notify-scripts.adoc (hunk 3 above)
Comment (ahardin-rh) [reply modules/nw-ipfailover-configuring-check-notify-scripts.adoc:12]: Thank you! I'm going to use " `ConfigMap`object" here per https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#object-references and update any other instances to align with this usage.

File: modules/nw-ipfailover-virtual-ip-addresses-concept.adoc (hunk 10 above)
Comment (ahardin-rh) [reply modules/nw-ipfailover-virtual-ip-addresses-concept.adoc:6]: Oops! I intended to delete this after reworking the content. Thanks for catching this!

File: networking/configuring-ipfailover.adoc (hunk 19 above)
Comment (ahardin-rh) [reply networking/configuring-ipfailover.adoc:10]: Renamed the child module. Thanks!

**Comment by rfredette** [comment]
Looks good to me!

**Pull request merged**